add gather to ShardedTensor

dracifer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65671



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

dracifer committed:
add gather to ShardedTensor

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -456,6 +456,35 @@ def test_create_sharded_tensor_with_ones(self):
+        _sharded_tensor.gather(sharded_tensor, 0, full_tensor)
+
+        if self.rank == 0:
+            self.assertEqual(full_tensor, torch.ones(h, 2))
```
Shouldn't this be `torch.ones(h, w)`? Where does the `2` come from?

dracifer replied:
typo...

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +352,54 @@ def _init_rpc(self):
+                    )
+                dims = len(sizes)
+                pad = [0] * 2 * dims
+                for dim in range(dims):
```
Are we padding each individual shard to be the size of the actual tensor and then just adding it? This is probably wasteful from a memory perspective. Can't we just get a view of `dst_tensor` based on offsets and lengths and just copy the shard into the appropriate position? You can probably use narrow (https://pytorch.org/docs/stable/generated/torch.narrow.html) multiple times to get this?

dracifer replied:
I discussed this as well in the internal diff. let me know your thoughts

dracifer commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -495,3 +495,23 @@ def shard_parameter(
+    Returns:
+        None
+    """
+    return ShardedTensor._gather(sharded_tensor, dst_rank, dst_tensor)
```
to confirm, we can remove this wrapper from __init__? curious what is the rule of thumb about when to put a wrapper in init and when not? thanks!

pritamdamania87 replied:
Most of the methods in the init file are around creating a ShardedTensor and not necessarily methods that modify an existing ShardedTensor. If something acts on a instance of ShardedTensor its probably just have it as part of the ShardedTensor class instead of having another wrapper.

dracifer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +352,54 @@ def _init_rpc(self):
+
+        gathered_shards = [None for _ in sharded_tensor._process_group.world_size]
+
+        dist.gather_object(
```
I commented in the internal diff on the same topic. was wonder about the performance overhead here and also wonder about the tradeoff if we flatten locally (larger payload to communicate and more memories). so the local flattening is better?

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -456,6 +456,35 @@ def test_create_sharded_tensor_with_ones(self):
+    def test_gather_sharded_ones(self):
+        """ Test _sharded_tensor.ones(...) """
+
+        spec = ChunkShardingSpec(
```
added

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -176,6 +177,22 @@ class TensorInitParams(object):
+            raise ValueError(
+                "Argument ``dst_tensor`` must be specified on destination rank."
+            )
+        if torch.count_nonzero(dst_tensor):
```
I second @pritamdamania here, we should be able to overwrite any tensor by filling it with the gathered content, there might be cases that user want to use a random tensor and fill it with the gathered content (although this might be rare).

dracifer replied:
got it. I will remove this check then. but just to ensure my understanding correctly, I will not zero the tensor in gather() and add the shards to it directly

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
+        my_rank = dist.get_rank(self._process_group)
+        full_size = self.metadata().size
+        _validate_output_tensor_for_gather(my_rank, dst, full_size, out)
+        curr_device = torch.device(f"cuda:{my_rank}")
```
why we need to use `curr_device` here?

pritamdamania87 replied:
As mentioned above, this won't work for ranks > 8

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
+                assert shards is not None, f"gathered shards cannot be None on {dst}"
+                for shard in shards:
+                    metadata = shard.metadata
+                    tensor = shard.tensor.to(curr_device)
```
there might be case that the tensors gathered together can't fit into a single GPU device right (that's the main reason user might want to shard the tensor). If the whole point of this `gather` call is to gather the content and fill the passed in tensor, should we not care about the device and just copy the content to the `out`?

dracifer replied:
I think it is a valid concern about size. @wanchaol so we can just do copy_ as you mentioned below. is it correct?

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
+                            metadata.shard_lengths[dim],
+                        ).contiguous()
+
+                    out_narrow_view.add_(tensor)
```
I guess the main reason you are asserting `zero` for `out` is that you are adding the gathered content to `out`, I think ideally we can just `copy_` it.

dracifer replied:
I see. thanks

wanchaol reviewed (commented):
@dracifer Thanks for the work! I did an initial pass of review, looks mostly good, left some questions and suggestions in the inlined comments.

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1187,6 +1220,53 @@ def test_create_sharded_tensor_with_ones(self):
+    def test_gather(self) -> None:
+        """ Test _sharded_tensor.gather(...) """
+
+        spec = EnumerableShardingSpec([
```
that is the only different. the two tests (all named test_gather) are testing the same function for different sharding strategies so make sure if works for both. they are implemented under different class TestShardedTensorEnumerable and TestShardedTensorChunked, respectively

dracifer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
+        gathered_shards = [None] * world_size
+        # all_gather_object is not efficient. will revise this part once
+        # NCCL support for dist.gather() is ready
+        with torch.cuda.device(curr_device):
```
I may not have the full context about cpu support. added this as @pritamdamania87 pointed out in prev comment. https://fburl.com/midhqrje. also from test cases, they seems to target on nccl. https://github.com/pytorch/pytorch/blob/master/test/distributed/_sharded_tensor/test_sharded_tensor.py#L269 please let me know if gpu only is expected.

pritamdamania87 replied:
Lets keep the scope of this PR small to address only GPU and then have a follow up to enable CPU as well. Please create a gh issue with the label `sharded_tensor` so we can keep track of this.

pritamdamania87 replied:
Instead of `curr_device` just use the device for the `out` tensor here.

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -194,7 +222,7 @@ class ShardedTensor(object):
     torch.empty
 
-    Args:
+    Arg:
```
nit still Args since its multiple?

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
+        _validate_output_tensor_for_gather(my_rank, dst, full_size, out)
+        curr_device = torch.device(f"cuda:{my_rank}")
+
+        shard_tensors = self.local_shards()
```
nit: local_shards

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
+            out (:class `torch.Tensor` optional): The output full tensor. Must to be provided
+            ONLY on `dst` rank. Default is None.
+        """
+        my_rank = dist.get_rank(self._process_group)
```
nit rank

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
+
+        gathered_shards = [None] * world_size
+        # TODO all_gather_object is not efficient. will revise this part once
+        # NCCL support for dist.gather() is ready
```
is below with command necessary?

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
+        the same `dst`. `out` should be a full size zero tensor on `dst` and
+        None on all other ranks.
+
+        TODO: current only supports GPU. Will enable CPU and revise to replace
```
nit: probably add a warning instead?

pritamdamania87 replied:
don't put todo in public docs, put this as a comment in the code and in the comment link to the appropriate gh issue.

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
+        if my_rank == dst:
+            dims = len(full_size)
+            for shards in gathered_shards:
+                assert shards is not None, f"gathered shards cannot be None on {dst}"
```
nit maybe raise exception instead of using assert to make it consistent?

pritamdamania87 replied:
Yes, this should raise a `RuntimeError`.

bowangbj reviewed (commented):
Thanks Bin, looks pretty good, just a few small styling issues

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
+        Creates a full :class:`Tensor` on rank `dst` by gathering all sharded tensors.
+
+        The API needs to be called on all ranks in SPMD fashion. All ranks should have
+        the same `dst`. `out` should be a full size zero tensor on `dst` and
```
nit: Out doesn't need to be a zero tensor, rephrase this to "out should be a tensor of the same size as the overall size of the sharded tensor".

dracifer replied:
good catch. this is a leftover from previous revision. will update

dracifer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -176,6 +177,33 @@ class TensorInitParams(object):
+            raise ValueError(
+                f"Argument ``dst_tensor`` must be specified on destination rank {dst_rank}"
+            )
+        curr_device = torch.device(f"cuda:{my_rank}")
```
right. I was a bit skeptical for this as well. any API suggestion here?

dracifer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -176,6 +177,33 @@ class TensorInitParams(object):
+                f"Argument ``dst_tensor`` have size {tuple(dst_tensor.size())},"
+                f"but should be {tuple(size)}"
+            )
+        if dst_tensor.device != curr_device:
```
this is mainly checking if dst_tensor is only created on dst_rank as we ask the user to provide dst_tensor only on it.

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]

bowangbj reviewed (commented):
Looks great for me. Please also wait for Pritam/Wanchao's cmt.

pritamdamania87 reviewed (approved):
LGTM, just one additional comment regarding more tests.

dracifer committed:
Update on "add gather to ShardedTensor"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

[ghstack-poisoned]
